man/man3/queues.3
.TH QUEUES 3X "August 6, 1990"
.SH NAME
queues \- tile kernel double linked list extensions
.SH SYNOPSIS
.B "#include queues.f83"
.LP
.B "queues"
.SH DESCRIPTION
The 
.IR tile
forth kernel word set for support of double linked circular lists.
Used by the kernel to maintain the run-time queue of tasks for
the multi-tasking extension.
.TP
.B
: .queue ( queue -- )
Displays information about a queue header or item in the format:
.br
.B queue#
.I <queue-pointer>
.B succ:
.I <succ-pointer>
.B pred:
.I <pred-pointer>
.br
.TP
.B
code ?empty-queue ( queue -- bool)
Returns "true" if the queue is empty else "false". Successor pointer
points to the queue itself. This definition is also available as forth
code for portability to other environments.
.TP
.B
: ?map-queue ( queue block[item -- bool] -- )
Used in the following form:
.br
.I <queue> 
.I <conditional-block> 
.B ?map-queue
.br
to perform a block of code on each element of the queue until
the block returns "true".
.TP
.B
: ?member-queue ( item queue -- bool)
Search for "item" in "queue". If found returns "true" else
"false". 
.TP
.B
struct.type QUEUE ( -- )
Structure definition of double linked list. Used as follows 
to create a new queue header instance:
.br
.B new-struct QUEUE ( -- queue)
.br
A named queue header may be created with:
.br
.B QUEUE 
.I <queue-name>
.B ( -- queue)
.br
Contains two private fields, "+succ" and "+pred".
.TP
.B
code dequeue ( item -- )
Removes "item" from any queue. Observe this function does not 
require knowledge about which queue "item" is a member of.
This definition is also available as a colon definition for
portability of other environments.
.TP
.B
code enqueue ( item queue -- )
Inserts "item" into "queue" as the new predecessor element. If
the "queue" parameter is a queue header "item" is inserted last 
into the queue. "item" must be a pointer to a "QUEUE" field of a 
structure or a sub-structure of "QUEUE".
This definition is also available as a colon definition for
portability of other environments.
.TP
.B
: map-queue ( queue block[item -- ] -- )
Used in the following form:
.br
.I <queue> 
.I <block> 
.B map
.br
Calls the parameter code "block" on each item in the queue. 
.TP
.B
vocabulary queues ( -- )
Vocabulary containing double linked circular list extensions.
Include into the vocabulary search structure, "context", to 
gain access to these extensions.
.TP
.B
: size-queue ( queue -- num)
Returns the length of a queue. The queue head is counted. 
Returns a integer larger than zero.
.TP
.B
: succ ( queue1 -- queue2) 
Returns pointer to successor queue item.
.TP
.B
: pred ( queue1 -- queue2)
Returns pointer to predecessor queue item.
.SH INTERNALS
Private definitions in the 
.I queues
vocabulary;
.TP
.B 
ptr +succ ( queue -- addr) private
Field of the structure type "QUEUE". Modifiers a queue pointer
to access successor pointer in queue structure. 
.TP
.B
ptr +pred ( queue -- addr) private
Field of the structure type "QUEUE". Modifiers a queue pointer
to access predecessor pointer in queue structure.
.SH "SEE ALSO"
.IR tile(1),
.IR forth(3X),
.IR structures(3X),
.IR blocks(3X).
.\" .SH EXAMPLES
.SH NOTE
The function list is sorted in ASCII order. 
The type and mode of the entries are indicated together with their
parameter stack effect.
.\" .SH WARNING
.\" .SH BUGS
.SH COPYING
Copyright (C) 1990 Mikael R.K. Patel
.PP
Permission is granted to make and distribute verbatim copies
of this manual provided the copyright notice and this permission
notice are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions
of this manual under the conditions for verbatim copying, 
provided also that the section entitled "GNU General Public
License" is included exactly as in the original, and provided
that the entire resulting derived work is distributed under
the terms of a permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of
this manual into another language, under the above conditions
for modified versions, except that the section entitled "GNU
General Public License" may be included in a translation approved
by the author instead of in the original English.
.SH AUTHOR
.nf
Mikael R.K. Patel
Computer Aided Design Laboratory (CADLAB)
Department of Computer and Information Science
Linkoping University
S-581 83 LINKOPING
SWEDEN
Email: mip@ida.liu.se
.if


src/queues.v
/*
  C BASED FORTH-83 MULTI-TASKING KERNEL: DOUBLE LINKED LIST

  Copyright (C) 1990 by Mikael R.K. Patel

  Computer Aided Design Laboratory (CADLAB)
  Department of Computer and Information Science
  Linkoping University
  S-581 83 LINKOPING
  SWEDEN

  Email: mip@ida.liu.se

  Started on: 12 April 1990

  Last updated on: 26 June 1990

  Dependencies:
	(cc) kernel.c, kernel.h

  Description:
	Double linked list (queues) extension vocabulary for the tile
	forth multi-tasking kernel.

  Copying:
       This program is free software; you can redistribute it and/or modify
       it under the terms of the GNU General Public License as published by
       the Free Software Foundation; either version 1, or (at your option)
       any later version.

       This program is distributed in the hope that it will be useful,
       but WITHOUT ANY WARRANTY; without even the implied warranty of
       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
       GNU General Public License for more details.

       You should have received a copy of the GNU General Public License
       along with this program; see the file COPYING.  If not, write to
       the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. 

*/

VOID doqemptyqueue()
{
    compare(== ((INT32) tos.QUEUE -> succ), INT32);
}

NORMAL_CODE(qemptyqueue, forth, "?empty-queue", doqemptyqueue);

VOID doenqueue()
{
    register QUEUE t, q;

    q = spop(QUEUE);
    t = spop(QUEUE);

    t -> pred = q -> pred;
    t -> succ = q;

    q -> pred -> succ = t;
    q -> pred = t;
}

NORMAL_CODE(enqueue, qemptyqueue, "enqueue", doenqueue);

VOID dodequeue()
{
    register QUEUE t;

    t = spop(QUEUE);

    t -> succ -> pred = t -> pred;
    t -> pred -> succ = t -> succ;

    t -> succ = t -> pred = t;
}

NORMAL_CODE(dequeue, enqueue, "dequeue", dodequeue);

